### ISSUE TITLE
"List Fold (Sequential)" Haskell benchmark isn't entirely faithful

### ISSUE THREAD

**Issue by sgraf812** [Kindelia/HVM#72]
I appreciate the work and thought you put into this project! It got me quite excited, to be honest; now I really want to know if/how interaction nets can actually be mapped to stock hardware efficiently.

But it appears to me that the benchmarks comparing to GHC-compiled Haskell aren't *entirely* faithful. You say 

>Since lists are sequential, and since there are no higher-order lambdas, HVM doesn't have any technical advantage over GHC.

That's technically true, but it's also not the kind of code you'd write in Haskell. For one, the tail-recursive nature of `range` means that you really *have to* build up a list that has to be copied during GCs. But Haskell (and HVM) is a lazy language and I'd rather write a generative definition of `range`, like

```hs
range 0 = Nil
range n = Cons n (range (n-1))
```

And with this definition, the entire program operates in constant space (modulo GC) due to the perks of laziness. Here's the complete program:

```hs
import System.Environment
import Data.Word

data List a = Nil | Cons a (List a)

-- Folds over a list
fold Nil         c n = n
fold (Cons x xs) c n = c x (fold xs c n)

-- A list from 0 to n
range 0 = Nil
range n =
  let m = n - 1
  in Cons n (range m)

-- Sums a big list with fold
main = do
  n <- read.head <$> getArgs :: IO Word32
  let size = 1000000 * n
  let list = range size
  print $ fold list (+) 0
```

This program takes about 1s for `n=30`, the same as the original program takes for `n=10` for me (and copies 1.6GB during GC). So about a factor 3 speedup.

Interaction nets might enjoy similar speedups, but I don't know that for sure. So it seems like a bit of an apples to oranges comparison.

Ultimately, making use of `foldl'` and list fusion on `[0..size-1::Int]` (`Word32` is broken) means another factor 100 of a speedup, of which I similarly don't know how it carries over to interaction nets.

**Comment by VictorTaelin** [comment]
I disagree this isn't the kind of code you write in Haskell, in general. Yes, for the problem folding over a list, specifically, it is possible to optimize it to a constant-space loop... in some cases. But that doesn't apply to basically any other Haskell function, which rely a lot on deep recursion. Even fold itself sometimes can't be turned into a constant-space loop. For example, if it was a subtraction, we couldn't use `foldl`, because `-` isn't associative.

That said, purpose of this benchmark was to **stress-test how efficient each engine is at recursing, pattern-matching and allocating**. If you replace that by a function that does no recursion, then it wouldn't be benchmarking how well the engine recurses, but instead, how well it deals with tight loops. I've tested, and HVM performs about ~3x slower than GHC in this case, which I guess is expected, because HVM does nothing to optimize these tight loops... yet! But it absolutely could, in which case both obviously would perform the same, since they'd generate the same low-level code. 

Remember, **HVM is not optimized**! It is the first version of a prototype, compared to a mature state-of-art compiler. It should be much slower than GHC, since devs put decades of effort into optimizing it. For example, GHC has an arena-based allocator, which makes great use of cache. HVM doesn't. Yet, it *does* perform as well as GHC on recursive functions. That's actually surprising. Think of how well it would perform if it had a fraction of GHC's total men-hours dedicated into optimizing it? That's the message I'm trying to pass.

**Comment by sgraf812** [comment]
> Think of how well it would perform if it had a fraction of GHC's total men-hours dedicated into optimizing it?

Well, that is actually what I'm trying to find out. TBH, I could speed up the program by another factor of 10 if I made use of list fusion, entirely eliminating all sources of allocation. I wonder if these factors carry over to HVM, too, if someone put in the work to implement a few foldr/build fusion primitives (which shouldn't be too hard). It's hard to tell, because as you say it obviously hasn't seen as much careful work put into it.

But at the same time, you can write programs in basically *any* language and instantly be in the realm of "a factor of 30 within the perf of C/GHC/whatever". The question is if you can ultimately deliver on the remaining constant factors. Pretending like those constant factors don't exist and say "we can do the same later on" is quite hand-wavy. Plus, speedups in GHC Haskell that are simply due to less GC pressure can't possibly carry over to HVM, because it doesn't do any GC (although it does a variant of reference counting, so that might count as GC after all). So it's a bit unproductive to do the comparison with a program that has a huge and continually growing memory residency, which is kind of a worst case for non-incremental, generational GC and still claim that the program is a workload that GHC should excel at.

I would much rather see "This is the most efficient way in GHC Haskell to solve this problem, and this is the most efficient way to solve it as an HVM program". And if GHC Haskell is 10 times faster in some cases, so be it! You can continually chip off the delta as you improve your VM implementation.

**Comment by VictorTaelin** [comment]
> Well, that is actually what I'm trying to find out. TBH, I could speed up the program by another factor of 10 if I made use of list fusion, entirely eliminating all sources of allocation. I wonder if these factors carry over to HVM, too, if someone put in the work to implement a few foldr/build fusion primitives (which shouldn't be too hard). It's hard to tell, because as you say it obviously hasn't seen as much careful work put into it.

Note that compile-time foldr/fusion primitives are **not** required to have fusion in HVM. It happens at runtime, naturally. Now, compile-time inlining is a different thing, and, yes, aggressive compile-time inlining can result in faster programs, essentially because you moved work to compile time. And, yes, HVM, once mature enough, could produce a `foldl'` add function that is exactly equivalent to the one produced by Haskell. There is no reason not to. Simple proof: compile HVM to Haskell, compile Haskell to Core (using GHC), compile Core back to HVM. We'll have the exact same inlined function as GHC, and we'll be free to compile it to a loop as well. That is why benchmarking foldl' is kinda pointless, since GHC does a lot of compile-time inlining, which HVM won't. It isn't a compiler, it is just the runtime. And in fact, compiling Haskell to HVM via GHC is one of the possible ways to use it!

> I would much rather see "This is the most efficient way in GHC Haskell to solve this problem, and this is the most efficient way to solve it as an HVM program". 

I think you do not realize though that HVM operates basically identically to GHC until there are high-order lambdas involved. For programs like `sum (range 0 n)`, optimal evaluation and lazy evaluation with thunks are the same... because they aren't even used! Since this program has no need for runtime closures, it just compiles to a C loop anyway. Whatever GHC can do, HVM can do.

Does that make sense?

**Comment by sgraf812** [comment]
> What I'm saying is: HVM is not an optimizer, it is just the runtime. So, if you want a fair comparison, it is STG vs HVM, not GHC vs HVM. A (kinda) reasonable way to do it is to compile Haskell with GHC, and then run the resulting program on both GHC and HVM. Are we on the same page here?

Yes, I agree with that. I'd even be fine with looking at the STG code, identifying some code patterns that you wouldn't generate if HVM was the target and replace them with more perf-aware patterns, playing the role of an HVM-backend of GHC.

With that in mind, your comparison isn't entirely faithful again, because you are using `Word32` instead of `Int`, which, as I said in the OP, has a broken `Enum` instance that doesn't allow fusion:

> Ultimately, making use of foldl' and list fusion on [0..size-1::Int] (Word32 is broken) means another factor 100 of a speedup

For `Word32` and `Int32` you get STG that is very different to the `Go` function I gave.

Fortunately, the issue is already fixed on GHC master (and I believe will also be fixed in GHC 9.0.3 and 9.2.2). So there are multiple ways in which you can arrive at good code:

1. Use GHC master
2. Replace `Word32` with `Word` (which seems to be even faster than `Int`)
3. Replace `Word32` with `Int`

So, for example

```hs
import Data.List
import Data.Word

-- Sums a big list with fold
main = do
  let size = 100000000 :: Word
  let list = [0..size-1]
  print $ foldl' (+) 0 list
```

executes in 50ms on my machine. 

Heck, we can smiply skip the whole list fusion shenigans and just benchmark a counting loop in Haskell, too. I'll simply translate your `Go` function back to Haskell:

```hs
import Data.Word

go :: Word32 -> Word32 -> Word32 -> Word32
go size i s = aux (i == size) size i s
  where
    aux True  size i s = s
    aux False size i s = go size (i + 1) (s + i)

main = do
  print $ go 100000000 0 0
```

Which takes 72ms on 8.10.7 (which has pretty bad sized Word support because of recently added sized primops) and 26ms on GHC master (which is the same perf as with `Word`).

If I want to evaluate this 8 times, like

```hs
import Data.Word

go :: Word32 -> Word32 -> Word32 -> Word32
go size i s = aux (i == size) size i s
  where
    aux True  size i s = s
    aux False size i s = go size (i + 1) (s + i)

main = do
  print $ go 100000000 0 0
  print $ go 100000000 0 1
  print $ go 100000000 0 2
  print $ go 100000000 0 3
  print $ go 100000000 0 4
  print $ go 100000000 0 5
  print $ go 100000000 0 6
  print $ go 100000000 0 7
```

it takes 580ms on 8.10.7 and 200ms on GHC master. That's way faster still than your parallelised program. Now you say

> I wouldn't even care if it was 10x slower in single-thread, no-allocations, non-optimal scenarios, like that one. HVM is cool because it allows automatic parallelism, no garbage-collection and beta-optimality.

If you want to focus on auto parallelisation, you might like https://wiki.haskell.org/GHC/Data_Parallel_Haskell where they tried something very similar. Meanwhile, all the DPH stuff had been removed from GHC again, because nobody was there who used it or maintained it. As always, the cases where you can't make use of your cores through (superior, because requires less sync) data/task paralellism are indeed rare and at some point the synchronisation overhead dominates. I wouldn't expect HVM to reach 200ms on this benchmark even with 80 cores, simply because of contention issues. So "50x slower" indeed matters.

I don't buy the No GC claim, because effectively the `dup` nodes take the role of reference counters (generalised, of course!) and that is a form of GC.

Beta optimality is indeed a nice and flat win. But I wonder on what workloads it really matters compared to STG-compiled programs that GHC produces (e.g., making use of transformations such as FloatOut/full laziness, simplification, eta expansion, ...). And obviously, I wonder what the costs are that pay for the win. It seems there are quite severe losses for sequential code.
**This is actually the question that matters most to me: Is Beta optimality worth the losses? Will the losses get significantly smaller over time?**

> That said, I'm not sure why you find this benchmark useful at all. It does no allocations, but allocations and garbage collection play a major role on functional runtimes. I don't know of any Haskell program that doesn't do a lot of allocation, so, this example isn't representative of real-world usage.

That allocations play a major role in functional programs is true; but we have to compare apples to apples somehow, right?

And it seems we just established that a simple counting loop will regress by a factor of ~50 when switching to a hypothetical HVM backend for GHC (as implemented at the moment! Not saying that won't change. But I still wonder how). This is the kind of hard and honest data I'm interested in! At the same time, there are cases and programs where GHC performs notoriously bad because of churn on the GC during build up of huge data structures. In these cases, HVM seems to be nearly on par. Your main README makes it sound like that is the common case for functional programs. I claim it is not; at some point the data structure is built up and ends up in Gen 1, where we will seldomly trace it during Gen 0 GC.

Besides, Haskell often has "fast non-allocating hot loops" such as repeated `(Int)Map.lookup` over the same map. And I think due to the `dup`ping that must be involved (e.g., for N lookups over the same map, the root would have to be `dup`ped N times, correct?), that'd suffer quite badly when compiled to HVM.

Indeed, all benchmarks you give build up a data structure and then consume it exactly once. Maybe compare examples where the same list or tree is traversed N times and benchmark that.

I stand by my initial claim that the README suggests a false balance, namely that HVM 

1. is almost on par with GHC's (STG to native code) backend for sequential workloads
2. faster when autoparallelisation is on 
3. and even faster when it manages to make an asymptotic difference due to beta optimality

(2) doesn't excite me very much TBH and we can have it (had it) in Haskell, too (others might be excited by it), (3) is nice to have, just as list fusion, and I think (1) is simply not true (at the moment!), as the counting loop example demonstrates.

**Comment by VictorTaelin** [comment]
Hello again! I missed this comment. To answer (if a little late):

> This is actually the question that matters most to me: Is Beta optimality worth the losses? Will the losses get significantly smaller over time?

My argument is that there are no losses. There is nothing on HVM that makes it inherently slower than GHC. In some cases, today, it will be, because GHC has a constellation of micro-optimizations that HVM doesn't.  Sometimes GHC will go as far as compiling a long loop to a constant, so obviously it will be faster. But that's all. One is a 3-decades-old **compiler and runtime**, improved by major unis and companies, the other is a new **runtime** written by a single dev. If you want a fair comparison, take a Haskell program, compile to GHC Core, and run it on both the HVM and on GHC's runtime. You'll notice they're very close together.

> And it seems we just established that a simple counting loop will regress by a factor of ~50 when switching to a hypothetical HVM backend for GHC

No, that's not the case. Take GHC core and run it on HVM.

> Besides, Haskell often has "fast non-allocating hot loops" such as repeated (Int)Map.lookup over the same map. 

No, a suitable optimizer should replace that by a linear version, so there will be no dup at all. Kind should take care of that kind of situation.

> I stand by my initial claim that the README suggests a false balance, namely that HVM
> 1. is almost on par with GHC's (STG to native code) backend for sequential workloads
> 2. faster when autoparallelisation is on
> 3. and even faster when it manages to make an asymptotic difference due to beta optimality

I do not suggest that. My claim is that HVM is getting closer and closer (but not almost on par) with GHC in general, and will be significantly faster than it soon (in some cases, already) for real applications due to parallelism and GC-freedom. If you see any sentence that suggests otherwise, please let me know.
